### ISSUE TITLE
Add Field for "Invoiced Object Identifier" (BT-18) in Invoice.java to Align with XRechnung Requirements

### ISSUE THREAD

**Issue by fktrdui** [ZUGFeRD/mustangproject#536]
The Invoice.java class currently lacks a field for storing the "Invoiced Object Identifier" (BT-18), which is required at the entire-invoice level in XRechnung. This identifier is essential for referencing objects related to the entire invoice, similar to the way Item.java stores "Invoice Line Object Identifier" (BT-128).

![BT18](https://github.com/user-attachments/assets/dcab1167-4a8d-4f29-b5b2-c22091ab2c04)

![BT128](https://github.com/user-attachments/assets/92521157-7ab1-43c7-b5c1-665b763befd3)

Mustang has the specification of mapping BT-18 to XML as follows:
(https://github.com/ZUGFeRD/mustangproject/blob/master/library/src/main/resources/stylesheets/cii-xr.xsl)
`<xsl:apply-templates mode="BT-18"                                 select="./rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeAgreement/ram:AdditionalReferencedDocument/ram:IssuerAssignedID[following-sibling::ram:TypeCode='130']"/>
`

In the current setup:
Item.java includes field (referencedDocuments) to store ReferencedDocument objects, allowing it to handle "Invoice Line Object Identifier" (BT-128).

However, Invoice.java has no corresponding field to hold ReferencedDocument instances for BT-18 at the invoice level.

**Suggested Solution:** 
To align with XRechnung requirements, add an ArrayList<ReferencedDocument> field in Invoice.java to store BT-18 references. This new field would ensure consistency across classes, allowing Invoice.java to manage invoice-wide object identifiers similarly to how Item.java handles line-specific identifiers.

```
public class Invoice {

    // New field for storing Invoiced Object Identifier (BT-18)
    protected ArrayList<ReferencedDocument> invoiceReferencedDocuments = null;

    // Getter for BT-18
    public ArrayList<ReferencedDocument> getInvoiceReferencedDocuments() {
        return invoiceReferencedDocuments;
    }

    // Setter for BT-18
    public void setInvoiceReferencedDocuments(ArrayList<ReferencedDocument> invoiceReferencedDocuments) {
        this.invoiceReferencedDocuments = invoiceReferencedDocuments;
    }

    // Method to add a single ReferencedDocument for BT-18
    public void addInvoiceReferencedDocument(ReferencedDocument doc) {
        if (invoiceReferencedDocuments == null) {
            invoiceReferencedDocuments = new ArrayList<>();
        }
        invoiceReferencedDocuments.add(doc);
    }
}

```

**Suggested XML Output Update:**
With the proposed changes, Invoice.java should generate XML for ./rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeAgreement/ram:AdditionalReferencedDocument
as follows:
```

if (trans.getInvoiceReferencedDocuments() != null) {
    for (final IReferencedDocument currentReferencedDocument : trans.getReferencedDocuments()) {
        xml += "<ram:AdditionalReferencedDocument>" +
               "<ram:IssuerAssignedID>" + XMLTools.encodeXML(currentReferencedDocument.getIssuerAssignedID()) + "</ram:IssuerAssignedID>" +
               "<ram:TypeCode>" + XMLTools.encodeXML(currentReferencedDocument.getTypeCode()) + "</ram:TypeCode>" +
               "<ram:ReferenceTypeCode>" + XMLTools.encodeXML(currentReferencedDocument.getReferenceTypeCode()) + "</ram:ReferenceTypeCode>" +
               "</ram:AdditionalReferencedDocument>";
    }
}
```

**Comment by johamb** [comment]
+1
This would also enable adding "Tender or lot reference" (BT-17).

**Comment by HoffmannTom** [comment]
InvoiceReferencedDocument is also needed in the final invoice to reference the partial invoices.

**Comment by fktrdui** [comment]
@HoffmannTom I see what you mean about InvoiceReferencedDocument for BG-3, which is used for referencing other invoices or partial invoices. That part was already handled in the correct XML position in ZUGFeRD2PullProvider. The only issue was that the properties in the Invoice class weren’t defined as an array, but that’s now "fixed".

My main concern, though, is about BT-18 ("Invoiced Object Identifier"), which serves a different purpose. It needs to be written in XML with TypeCode "130" and placed under ApplicableHeaderTradeAgreement, not ApplicableHeaderTradeSettlement:

`<xsl:apply-templates mode="BT-18"                                 select="./rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeAgreement/ram:AdditionalReferencedDocument/ram:IssuerAssignedID[following-sibling::ram:TypeCode='130']"/>`

The implementation and XML structure should be the same as additionalReferences in the Item class, which handles BT-128 ("Invoice Line Object Identifier"), but applied at the Invoice level.

**Comment by HoffmannTom** [comment]
@fktrdui 
Currently, both use the class ReferencedDocument
When used in the element <InvoiceReferencedDocument> only the attributes IssuerAssignedID and FormattedIssueDateTime are used and are specified. The TypeCode is not used because in InvoiceReferencedDocument it is always an invoice reference.

![Image](https://github.com/user-attachments/assets/033b2152-4e69-4c71-a94e-76430e5e65b0)

So I currently see no overlapping or problems. I am not sure whether I understood your concerns right.

**Comment by fktrdui** [comment]
@HoffmannTom
We’re talking about two different things. InvoiceReferencedDocument is for referencing other invoices, but my issue is about BT-18 ("Invoiced Object Identifier"), which has different properties like Type Code and Scheme Identifier.

As I see it, the issue with InvoiceReferencedDocument is already being handled in another ticket (#628).

**Comment by HoffmannTom** [comment]
@fktrdui 
I think we are on the same page. BT-18 is used to reference all kinds of documents, like price lists, additional documents etc.
The special case with 130 is used to reference invoices.
BG-3 is only for referencing invoices, so imho it is just a special case

**Comment by fktrdui** [comment]
@HoffmannTom 
I don’t think so. The different interpretations may come from different specifications. In the XRechnung specification, there are many referencing BT items, most of which are already implemented in the Invoice class of the Mustang project.

As you can see in the image I attached in my first comment, the data type of BT-18 is "Identifier", not "Document Reference". Clearly, they are not semantically the same.
Moreover, according to the specification, BT-18 provides additional information about the invoiced item, a specific property of the item, assigned by the seller.

**Comment by fktrdui** [comment]
@HoffmannTom 
my concern here is Invoiced object identifier(BT-18), which is neither AdditionalReferencedDocument(BG-24) nor InvoiceReferencedDocument(BG-1).

**Comment by HoffmannTom** [comment]
@fktrdui 
Right, two different entities. Unfortunately, the same name within ZUGFeRD specification.

**Comment by fktrdui** [comment]
@HoffmannTom
I reviewed the ZUGFeRD definitions and implementations again to avoid any misunderstandings. As implemented in the Item class, IReferencedDocument can be used to store BT-18. Similarly, the Invoice class should have two separate properties to store IReferencedDocument objects, and they should be handled correctly in ZUGFeRD2PullProvider, which is currently not the case.

The issue with the current Invoice class implementation is that the new property:
`ArrayList<ReferencedDocument> invoiceReferencedDocuments = null;`
is commented as "New field for storing Invoiced Object Identifier (BG-3)", which is incorrect.

The issues:
1. BT-18 (Invoiced Object Identifier) is not related to BG-3.
2. In the previous version, the information now stored in invoiceReferencedDocuments was originally saved in these two properties:
      protected String invoiceReferencedDocumentID = null;
      protected Date invoiceReferencedIssueDate;
However, ZUGFeRD2PullProvider now writes this data to XML from invoiceReferencedDocuments instead. If this new field is used, the old properties are no longer needed.
3. The issue of saving BT-18 (Invoiced Object Identifier) remains unresolved.
